include/nbl/ui/CWindowWin32.h

```diff
@@ -39,97 +41,150 @@ class CWindowWin32 final : public IWindowWin32
+				style |= WS_VISIBLE;
 			}
 
-			// ? not sure about those below
-			style |= WS_CLIPCHILDREN;
-			style |= WS_CLIPSIBLINGS;
-		}
-		if (isMinimized())
-			style |= WS_MINIMIZE;
-		if (isMaximized())
-			style |= WS_MAXIMIZE;
-		// TODO:
-		// if (hasMouseCaptured())
-		// if (hasInputFocus())
-		// if (hasMouseFocus())
-		// if (isAlwaysOnTop())
-
-		AdjustWindowRect(&clientSize, style, FALSE);
-
-		const int32_t realWidth = clientSize.right - clientSize.left;
-		const int32_t realHeight = clientSize.bottom - clientSize.top;
-
-		int32_t windowLeft = (GetSystemMetrics(SM_CXSCREEN) - realWidth) / 2;
-		int32_t windowTop = (GetSystemMetrics(SM_CYSCREEN) - realHeight) / 2;
-
-		if (windowLeft < 0)
-			windowLeft = 0;
-		if (windowTop < 0)
-			windowTop = 0;	// make sure window menus are in screen on creation
-
-		if (isFullscreen())
-		{
-			windowLeft = 0;
-			windowTop = 0;
-		}
 
-		// create window
+			// TODO:
+			// if (hasMouseCaptured())
+			// if (hasInputFocus())
+			// if (hasMouseFocus())
 
-		m_native = CreateWindow(classname, __TEXT(""), style, windowLeft, windowTop,
-			realWidth, realHeight, NULL, NULL, hinstance, NULL);
-		m_width = realWidth;
-		m_height = realHeight;
+			AdjustWindowRect(&clientSize, style, FALSE);
 
-		if (!isHidden())
-			ShowWindow(m_native, SW_SHOWNORMAL);
-		UpdateWindow(m_native);
+			const int32_t realWidth = clientSize.right - clientSize.left;
+			const int32_t realHeight = clientSize.bottom - clientSize.top;
 
-		// fix ugly ATI driver bugs. Thanks to ariaci
-		// TODO still needed?
-		MoveWindow(m_native, windowLeft, windowTop, realWidth, realHeight, TRUE);
-    }
+			int32_t windowLeft = (GetSystemMetrics(SM_CXSCREEN) - realWidth) / 2;
+			int32_t windowTop = (GetSystemMetrics(SM_CYSCREEN) - realHeight) / 2;
+
+			if (windowLeft < 0)
+				windowLeft = 0;
+			if (windowTop < 0)
+				windowTop = 0;	// make sure window menus are in screen on creation
+
+			if (_flags & ECF_FULLSCREEN)
+			{
+				windowLeft = 0;
+				windowTop = 0;
+			}
+
+			// create window
+
+			 nativeWindow = CreateWindow(classname, __TEXT(""), style, windowLeft, windowTop,
+				realWidth, realHeight, NULL, NULL, hinstance, NULL);
+			if ((_flags & ECF_HIDDEN) == 0)
+				ShowWindow(nativeWindow, SW_SHOWNORMAL);
+			UpdateWindow(nativeWindow);
+
+			// fix ugly ATI driver bugs. Thanks to ariaci
+			// TODO still needed?
+			MoveWindow(nativeWindow, windowLeft, windowTop, realWidth, realHeight, TRUE);
+		}
 
-    native_handle_t m_native;
+		native_handle_t getNativeWindow() const
+		{
+			return nativeWindow;
+		}
+	private:
+		void init()
+		{
+			// All the stuff was already initialized in the ctor
+		}
+		void exit()
+		{
+			DestroyWindow(nativeWindow);
+		}
+		void work(base_t::lock_t& lock)
+		{
+			// @criss not sure why PeekMessage is better than GetMessage, since we'll be spining all the time here
+			// Is a spinlock a better solution than blocking a thread?
+			keepRunning = GetMessage(&message, nativeWindow, 0, 0);
```

Comment:
you want `NULL` instead of `nativeWindow` here. In order to gather events for all windows created by this thread (effectively all win32 windows created by Nabla)

Reply:
https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getmessage

> If hWnd is NULL, GetMessage retrieves messages for any window that belongs to the current thread

Reply:
oh nice, now I can see how this could be made to have only 1 thread for all windows